Stream view

zeeres opened the pull request:
Adds a stream view menu with the following features:
- Auto-scroll: When players get highlighted, by default scroll to the first unhighlighted player
- Auto-select: Select the last highlighted player
- Keep: Keep this amount of highlighted people visible when using the auto-scroll feature
- Switch panels: Switches the packet-container with the right-panel

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+        if (!$scope.autoscroll) return;
+        var i=0;
+        while (i < $scope.packet.length && Object.keys($scope.highlights).indexOf($scope.packet[i].profile) !== -1) i++;;
+        $scope.max_keep = i;
```
I'm not really a fan of this `max_keep`. It currently means you can't set it to anything when nobody's been drafted yet, and I assume you'll want to set the `keep` once before everything starts. I'd suggest removing the upper limit altogether.

zeeres replied:
I agree than in general you'd want to set it once before the draft and not change it later. But it may be useful to have during draft when you want to scroll up/down a player. Not really necessary though.
I've introduced `max_keep` because
- you'll get an error when trying to scroll to the -1'th player (`#player--1`)
- to signal the user that it doesn't make sense to further increase the `keep` value
I could just not execute the scroll if the `keep` value if it doesn't make sense.

arfie replied:
Yeah, that was my thinking. Check if the row to scroll to is in range of the packet, and only scroll if it is.

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+        var i=0;
+        while (i < $scope.packet.length && Object.keys($scope.highlights).indexOf($scope.packet[i].profile) !== -1) i++;;
+        $scope.max_keep = i;
+        console.log(i);
```
Clean up your debug code! :)

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+    };
+    $scope.autoScroll = function() {
+        $('#packet-container').popover('hide');
+        $scope.autoscroll = $scope.autoscroll === false ? true : false;
```
`$scope.autoscroll = !$scope.autoscroll;` might be a little more readable.

zeeres replied:
haha, sometimes you do stupid things :D

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+        var keys = Object.keys($scope.highlights),
+            lastplayer = keys[keys.length-1],
+            i=0;
+        while ((i < $scope.packet.length) && ($scope.packet[i].profile !== lastplayer)) i++;;
```
I'm not 100% sure that `Object.keys` preserves order of when the keys were added. That probably depends on the engine. I assume most engines would use a hashtable to implement `Object`s, which would mean order is NOT preserved. Implementing this without using that would be tricky though. What do you think about syncing the `selection` through localstorage like we do with the highlights? And then getting rid of the auto select feature. Then you could also select other players at will (from your other window), instead of the ones that were just drafted.

zeeres replied:
Alright, don't know enough about engines. I've only been testing in FF and order was preserved there. I assumed it was a "feature" of javascript and the same across browsers.

Your suggestion to use `selection` for auto selecting seems good, I'll try it out!

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+        if ($('#packet-container').nextAll().filter('#right-panel').length !== 0)  // if right-panel is right of the packet-container
+            $('#right-panel').insertBefore('#packet-container');
+        else // right-panel is left
+            $('#right-panel').insertAfter('#packet-container');
```
Please please please don't use jquery to move elements. That is extremely inefficient and prone to break other stuff. I suggest you simply toggle a class on `#packet-container`, which changes its CSS `order` attribute. Long story short, setting the `order` to 2 (I think) will put it on the right of `#right-panel`, despite appearing earlier in the DOM.

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+        $('#keep').text(0);
+    };
+    $scope.changeKeep = function(i) {
+        if ((i < 0 && $scope.keep > 0) || i > 0 && $scope.keep < $scope.max_keep) {
```
That's a very complex predicate. Wouldn't `0 <= $scope.keep + i && $scope.keep + i <= $scope.max_keep` do the same? Either way, if you follow my earlier comment and get rid of `max_keep` you can simplify it a lot anyway.

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+    $scope.changeKeep = function(i) {
+        if ((i < 0 && $scope.keep > 0) || i > 0 && $scope.keep < $scope.max_keep) {
+            $scope.keep += i;
+            $('#keep').text($scope.keep);
```
Use Angular to display data, so you'd just put `{{ keep }}` in `#keep` and it will update automagically.

arfie commented on web/app.js:
```diff
@@ -341,6 +362,60 @@ app.controller('draftPacketController', function($scope, $http) {
+        if ((i < 0 && $scope.keep > 0) || i > 0 && $scope.keep < $scope.max_keep) {
+            $scope.keep += i;
+            $('#keep').text($scope.keep);
+            if ($scope.autoscroll)
```
This check is redundant as it also happens in the `auto_scroll` method.

arfie commented on web/index.html:
```diff
@@ -183,6 +184,30 @@
+            <a href="#" title="Keep player +1" data-toggle="tooltip" data-placement="right" ng-click="changeKeep(1)">
+                <i class="fas fa-plus"></i>
+            </a>
+            <p id="keep" title="Keep unhighlighted" data-toggle="tooltip" data-placement="right">0</p>
```
I'm not really a fan of this UI with the separate increase/decrease buttons. Wouldn't just having a text/number input (that looks like the other buttons) with `ng-model="keep"` be a lot cleaner?

zeeres replied:
Can you elaborate on this, how would the new button look like?

arfie replied:
The same as the number `p` does now, except you can input a number in it like in a textbox.

arfie commented on web/index.html:
```diff
@@ -183,6 +184,30 @@
+            </a>
+            <a href="#" title="Switch panels" data-toggle="tooltip" data-placement="right" ng-click="switchPanels()">
+                <i class="fas fa-columns"></i>
+            </a>
```
This is also really pedantic, but I'd like if the reset button is pushed to the bottom like on the other panels. So you'd put a `div.flex-filler` inbetween.

zeeres replied:
Not pedantic at all, it should be consistent!

arfie commented on web/style.css:
```diff
@@ -176,6 +176,10 @@ thead tr th {
 }
 
+.highlighter-white::after, .hl-white {
+    background-image: repeating-linear-gradient(45deg, white 0%, white 2%, rgba(235,235,235,0) 2%, rgba(235,235,235,0) 4%, white 4%) !important;
```
Oh and also, maybe "white" isn't a great name for this particular highlight since, well, it's not exactly white :P Though don't worry about that really, since we're gonna remove the highlight again when the draft updating is in.

zeeres replied:
Why would you remove it? How would you deal with mascot picks and auto-scroll?

No idea why it doesn't show on even rows, will try to fix it or do something else. Maybe just completely white like I originally intended, hence the name.

arfie replied:
I want to use the effect you used here for the players that are drafted, because it looks really cool. When I have that done, we can update the auto scroll to be executed when a new draft is done.

arfie replied:
And having it as a highlight as well would be kinda redundant, especially as there are already enough highlight options I think.

zeeres commented:
fixed it, how is it looking now?

zeeres commented:
Alright, it worked in Firefox but not Chromium for me. Apparently Chromium doesn't support smooth behavior for `scrollIntoView`. Removed it and added `block: "start"` to make Chromium scroll to the top of the element. That works in both browsers as far as I can tell. Can you test this on your PC?

arfie commented:
LGTM! If you pull in my recent changes and remove your drafted highlight (as it's built in now) it's good to merge.
